### ISSUE TITLE
Race condition on filesystems with low-precision mtime

### ISSUE THREAD

**Issue by hobarrera** [pimutils/khal#576]
Following up on #575, I realized that there's a possible (although very unlikely) race condition on file-systems where precision is only in seconds. During the same second:

1. A file is written.
1. The file is cached.
1. A file is re-written.

The cache will now keep the first version of the file, and will believe it's up-to-date.

As a possible solution, I'd like to propose:

* Measure the mtime precision for each calendar at startup (similar to [this](https://github.com/pimutils/khal/blob/62b015cb0befeff3634bf1d3927940ed3805b4be/tests/conftest.py#L55)).
* If the FS is lo-precision set some special internal flag to `slow_fs` to True.
* When reading files from disk if `mtime == int(datetime.now().timestamp()` (ie: the file was written during this second), set the mtime in the cache to *one second in the past*.

This will result in the cache for a file written during the second is was read *always* expiring. The frequency of this happening should be very very low, but we'll avoid the race condition outlined above.

To summarize, this would add _very very_ few occasional re-read, to avoid keeping stale data in the cache.

Rather than an actual PR, I wanted to discuss this first, because:

1. Will we care enough about this?
1. Maybe somebody has a better idea?

**Comment by untitaker** [comment]
If you really care about this you could also take a combination of inode number, file size and mtime. All of this is returned from `stat` at once so you don't have any extra syscalls

**Comment by hobarrera** [comment]
`inode` does not change across saves. I do agree that file size reduces this considerable, but I'm sure if it's enough.

**Comment by untitaker** [comment]
inode does change if you use atomic file writes.

**Comment by hobarrera** [comment]
Yes, you're right.
